### ISSUE TITLE
Replace which appends to string, executes twice

### ISSUE THREAD

**Issue by mindflow** [rollup/plugins#752]
- Rollup Plugin Name: plugin-replace
- Rollup Plugin Version: 2.3.4
- Rollup Version: 2.35.1
- Operating System (or Browser): Windows
- Node Version: 15.5.0
- Link to reproduction _(⚠️ read below)_: https://repl.it/join/ygbqcvmd-tobynilsen

### Expected Behavior

Expect the string to be replaced once.
"batman" -> "batman once"

### Actual Behavior

String is replaced twice,
"batman" -> "batman once once"

### Additional Information

I am assuming that the behavior of both renderChunk and transform leads to this:

    renderChunk(code, chunk) {
      const id = chunk.fileName;
      if (!keys.length) return null;
      if (!filter(id)) return null;
      return executeReplacement(code, id);
    },

    transform(code, id) {
      if (!keys.length) return null;
      if (!filter(id)) return null;
      return executeReplacement(code, id);
    }

**Comment by shellscape** [comment]
Hey thanks for the issue, but your reproduction link is broken

**Comment by mindflow** [comment]
Added rollup to the repl.it sharing now

**Comment by mindflow** [comment]
I have a proposal for a fix. I would ask if the options could include preferences about stage execution.

E.g
`{
'hello': 'hello world'
replacementStage : 'renderChunk'
}`

index.js

`import MagicString from 'magic-string';
import { createFilter } from '@rollup/pluginutils';

function escape(str) {
  return str.replace(/[-[\]/{}()*+?.\\^$|]/g, '\\$&');
}

function ensureFunction(functionOrValue) {
  if (typeof functionOrValue === 'function') return functionOrValue;
  return () => functionOrValue;
}

function longest(a, b) {
  return b.length - a.length;
}

function getReplacements(options) {
  if (options.values) {
    return Object.assign({}, options.values);
  }
  const values = Object.assign({}, options);
  delete values.delimiters;
  delete values.include;
  delete values.exclude;
  delete values.sourcemap;
  delete values.sourceMap;
  delete values.replaceStage;                                                                                                 // Added
  return values;
}

function mapToFunctions(object) {
  return Object.keys(object).reduce((fns, key) => {
    const functions = Object.assign({}, fns);
    functions[key] = ensureFunction(object[key]);
    return functions;
  }, {});
}

export default function replace(options = {}) {
  const filter = createFilter(options.include, options.exclude);
  const { delimiters, replaceStage } = options;                                                                  // Modified
  const functionValues = mapToFunctions(getReplacements(options));
  const keys = Object.keys(functionValues)
    .sort(longest)
    .map(escape);
  const pattern = delimiters
    ? new RegExp(`${escape(delimiters[0])}(${keys.join('|')})${escape(delimiters[1])}`, 'g')
    : new RegExp(`\\b(${keys.join('|')})\\b`, 'g');

  return {
    name: 'replace',

    renderChunk(code, chunk) {
      const id = chunk.fileName;
      if (replaceStage && replaceStage !== 'renderChunk') { return null; }                       // Added
      if (!keys.length) return null;
      if (!filter(id)) return null;
      return executeReplacement(code, id);
    },

    transform(code, id) {
      if (replaceStage && replaceStage !== 'transform') { return null; }                       // Added
      if (!keys.length) return null;
      if (!filter(id)) return null;
      return executeReplacement(code, id);
    }
  };

  function executeReplacement(code, id) {
    const magicString = new MagicString(code);
    if (!codeHasReplacements(code, id, magicString)) {
      return null;
    }

    const result = { code: magicString.toString() };
    if (isSourceMapEnabled()) {
      result.map = magicString.generateMap({ hires: true });
    }
    return result;
  }

  function codeHasReplacements(code, id, magicString) {
    let result = false;
    let match;

    // eslint-disable-next-line no-cond-assign
    while ((match = pattern.exec(code))) {
      result = true;

      const start = match.index;
      const end = start + match[0].length;
      const replacement = String(functionValues[match[1]](id));
      magicString.overwrite(start, end, replacement);
    }
    return result;
  }

  function isSourceMapEnabled() {
    return options.sourceMap !== false && options.sourcemap !== false;
  }
}
`

**Comment by shellscape** [comment]
IMHO that would place an undue burden on users to understand the Rollup lifecycle. Word boundaries typically handle your edge case: https://github.com/rollup/plugins/tree/master/packages/replace#word-boundaries. If we.change the plugin, it'll have to be with ease of use as a priority.

**Comment by mindflow** [comment]
@shellscape , I get your point. The default behavior of the plugin with the minimal config should be what makes the most sense. I am not sure that defining boundaries in order to prevent the side effect of having two stages which replaces each match, is optimal in terms of user friendliness, Nor am I sure that it would capture all problem cases. I had to dig deep to understand why I got the results I got. For now I am making my own custom plugin to get what I need.
